### PR TITLE
[dask] pass additional predict() parameters through when input is a Dask Array

### PR DESCRIPTION
**Pull request by jameslamb**
`.predict()` in the Dask estimators allows users to pass additional prediction parameters (https://lightgbm.readthedocs.io/en/latest/Parameters.html#predict-parameters) through `**kwargs`. To be applied correctly, those `**kwargs` have to be passed through several layers of function calls inside the package.

One of those pass-throughs is currently missing, and as a result additional prediction parameters will be silently ignored when data passed to `.predict()` is a Dask Array.

This PR proposes fixing that and adding tests confirming that additional parameters are being passed through correctly.

### Notes for Reviewers

I looked at `lightgbm.basic._InnerPredictor.predict()` for the names of specific parameters to be passed through.

https://github.com/microsoft/LightGBM/blob/bd21efed4f9dfffa5119775c835e1097d078085b/python-package/lightgbm/basic.py#L661-L663

### CODE REVIEW

File: tests/python_package_test/test_dask.py [hunk 1]
```diff
@@ -297,6 +299,10 @@ def test_classifier(output, task, boosting_type, tree_learner, cluster):
             assert_eq(p1_local, p2)
             assert_eq(p1_local, y)
 
+        # extra predict() parameters should be passed through correctly
+        with pytest.raises(AssertionError):
+            assert_eq(p1_raw, p1_first_iter)
```
Comment (jameslamb) [inline tests/python_package_test/test_dask.py:304]: on `master`, this test and the others added in this PR pass for all Dask Dataframe inputs and fail for all Dask Array inputs

output of `cd tests/python_package_test/ && pytest test_dask.py` on `master`:

```text
====================================== short test summary info ======================================
FAILED test_dask.py::test_classifier[data-gbdt-binary-classification-array] - Failed: DID NOT RAIS...
FAILED test_dask.py::test_classifier[data-gbdt-binary-classification-scipy_csr_matrix] - Failed: D...
FAILED test_dask.py::test_classifier[data-gbdt-multiclass-classification-array] - Failed: DID NOT ...
FAILED test_dask.py::test_classifier[data-gbdt-multiclass-classification-scipy_csr_matrix] - Faile...
FAILED test_dask.py::test_classifier[data-dart-binary-classification-array] - Failed: DID NOT RAIS...
FAILED test_dask.py::test_classifier[data-dart-binary-classification-scipy_csr_matrix] - Failed: D...
FAILED test_dask.py::test_classifier[data-dart-multiclass-classification-array] - Failed: DID NOT ...
FAILED test_dask.py::test_classifier[data-dart-multiclass-classification-scipy_csr_matrix] - Faile...
FAILED test_dask.py::test_classifier[data-goss-binary-classification-array] - Failed: DID NOT RAIS...
FAILED test_dask.py::test_classifier[data-goss-binary-classification-scipy_csr_matrix] - Failed: D...
FAILED test_dask.py::test_classifier[data-goss-multiclass-classification-array] - Failed: DID NOT ...
FAILED test_dask.py::test_classifier[data-goss-multiclass-classification-scipy_csr_matrix] - Faile...
FAILED test_dask.py::test_classifier[data-rf-binary-classification-array] - Failed: DID NOT RAISE ...
FAILED test_dask.py::test_classifier[data-rf-binary-classification-scipy_csr_matrix] - Failed: DID...
FAILED test_dask.py::test_classifier[data-rf-multiclass-classification-array] - Failed: DID NOT RA...
FAILED test_dask.py::test_classifier[data-rf-multiclass-classification-scipy_csr_matrix] - Failed:...
FAILED test_dask.py::test_classifier[voting-gbdt-binary-classification-array] - Failed: DID NOT RA...
FAILED test_dask.py::test_classifier[voting-gbdt-binary-classification-scipy_csr_matrix] - Failed:...
FAILED test_dask.py::test_classifier[voting-gbdt-multiclass-classification-array] - Failed: DID NO...
FAILED test_dask.py::test_classifier[voting-gbdt-multiclass-classification-scipy_csr_matrix] - Fai...
FAILED test_dask.py::test_classifier[voting-dart-binary-classification-array] - Failed: DID NOT RA...
FAILED test_dask.py::test_classifier[voting-dart-binary-classification-scipy_csr_matrix] - Failed:...
FAILED test_dask.py::test_classifier[voting-dart-multiclass-classification-array] - Failed: DID NO...
FAILED test_dask.py::test_classifier[voting-dart-multiclass-classification-scipy_csr_matrix] - Fai...
FAILED test_dask.py::test_classifier[voting-goss-binary-classification-array] - Failed: DID NOT RA...
FAILED test_dask.py::test_classifier[voting-goss-binary-classification-scipy_csr_matrix] - Failed:...
FAILED test_dask.py::test_classifier[voting-goss-multiclass-classification-array] - Failed: DID NO...
FAILED test_dask.py::test_classifier[voting-goss-multiclass-classification-scipy_csr_matrix] - Fai...
FAILED test_dask.py::test_classifier[voting-rf-binary-classification-array] - Failed: DID NOT RAIS...
FAILED test_dask.py::test_classifier[voting-rf-binary-classification-scipy_csr_matrix] - Failed: D...
FAILED test_dask.py::test_classifier[voting-rf-multiclass-classification-array] - Failed: DID NOT ...
FAILED test_dask.py::test_classifier[voting-rf-multiclass-classification-scipy_csr_matrix] - Faile...
FAILED test_dask.py::test_classifier_pred_contrib[multiclass-classification-scipy_csr_matrix] - At...
FAILED test_dask.py::test_regressor[data-gbdt-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_regressor[data-gbdt-scipy_csr_matrix] - Failed: DID NOT RAISE <class 'As...
FAILED test_dask.py::test_regressor[data-dart-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_regressor[data-dart-scipy_csr_matrix] - Failed: DID NOT RAISE <class 'As...
FAILED test_dask.py::test_regressor[data-goss-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_regressor[data-goss-scipy_csr_matrix] - Failed: DID NOT RAISE <class 'As...
FAILED test_dask.py::test_regressor[data-rf-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_regressor[data-rf-scipy_csr_matrix] - Failed: DID NOT RAISE <class 'Asse...
FAILED test_dask.py::test_regressor[voting-gbdt-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_regressor[voting-gbdt-scipy_csr_matrix] - Failed: DID NOT RAISE <class '...
FAILED test_dask.py::test_regressor[voting-dart-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_regressor[voting-dart-scipy_csr_matrix] - Failed: DID NOT RAISE <class '...
FAILED test_dask.py::test_regressor[voting-goss-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_regressor[voting-goss-scipy_csr_matrix] - Failed: DID NOT RAISE <class '...
FAILED test_dask.py::test_regressor[voting-rf-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_regressor[voting-rf-scipy_csr_matrix] - Failed: DID NOT RAISE <class 'As...
FAILED test_dask.py::test_ranker[data-gbdt-None-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_ranker[data-gbdt-group1-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[data-dart-None-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_ranker[data-dart-group1-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[data-goss-None-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_ranker[data-goss-group1-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[data-rf-None-array] - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED test_dask.py::test_ranker[data-rf-group1-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_ranker[voting-gbdt-None-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[voting-gbdt-group1-array] - Failed: DID NOT RAISE <class 'Asserti...
FAILED test_dask.py::test_ranker[voting-dart-None-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[voting-dart-group1-array] - Failed: DID NOT RAISE <class 'Asserti...
FAILED test_dask.py::test_ranker[voting-goss-None-array] - Failed: DID NOT RAISE <class 'Assertion...
FAILED test_dask.py::test_ranker[voting-goss-group1-array] - Failed: DID NOT RAISE <class 'Asserti...
FAILED test_dask.py::test_ranker[voting-rf-None-array] - Failed: DID NOT RAISE <class 'AssertionEr...
FAILED test_dask.py::test_ranker[voting-rf-group1-array] - Failed: DID NOT RAISE <class 'Assertion...
================ 65 failed, 214 passed, 3 skipped, 260 warnings in 120.86s (0:02:00) ================
```

File: tests/python_package_test/test_dask.py [hunk 2]
```diff
@@ -272,6 +272,8 @@ def test_classifier(output, task, boosting_type, tree_learner, cluster):
         )
         dask_classifier = dask_classifier.fit(dX, dy, sample_weight=dw)
         p1 = dask_classifier.predict(dX)
+        p1_raw = dask_classifier.predict(dX, raw_score=True).compute()
+        p1_first_iter = dask_classifier.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:276]:
```suggestion
        p1_first_iter_raw = dask_classifier.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```

File: tests/python_package_test/test_dask.py (hunk 1 above)
Comment (StrikerRUS) [reply tests/python_package_test/test_dask.py:304]:
```suggestion
            assert_eq(p1_raw, p1_first_iter_raw)
```

File: tests/python_package_test/test_dask.py [hunk 3]
```diff
@@ -487,6 +493,8 @@ def test_regressor(output, boosting_type, tree_learner, cluster):
 
         s1 = _r2_score(dy, p1)
         p1 = p1.compute()
+        p1_raw = dask_regressor.predict(dX, raw_score=True).compute()
+        p1_first_iter = dask_regressor.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:497]:
```suggestion
        p1_first_iter_raw = dask_regressor.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```

File: tests/python_package_test/test_dask.py [hunk 4]
```diff
@@ -516,6 +524,10 @@ def test_regressor(output, boosting_type, tree_learner, cluster):
         assert_eq(p1, y, rtol=0.5, atol=50.)
         assert_eq(p2, y, rtol=0.5, atol=50.)
 
+        # extra predict() parameters should be passed through correctly
+        with pytest.raises(AssertionError):
+            assert_eq(p1_raw, p1_first_iter)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:529]:
```suggestion
            assert_eq(p1_raw, p1_first_iter_raw)
```

File: tests/python_package_test/test_dask.py [hunk 5]
```diff
@@ -680,6 +692,8 @@ def test_ranker(output, group, boosting_type, tree_learner, cluster):
         rnkvec_dask = dask_ranker.predict(dX)
         rnkvec_dask = rnkvec_dask.compute()
         p1_pred_leaf = dask_ranker.predict(dX, pred_leaf=True)
+        p1_raw = dask_ranker.predict(dX, raw_score=True).compute()
+        p1_first_iter = dask_ranker.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:696]:
```suggestion
        p1_first_iter_raw = dask_ranker.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
```

File: tests/python_package_test/test_dask.py [hunk 6]
```diff
@@ -693,6 +707,10 @@ def test_ranker(output, group, boosting_type, tree_learner, cluster):
         assert spearmanr(rnkvec_dask, rnkvec_local).correlation > 0.8
         assert_eq(rnkvec_dask, rnkvec_dask_local)
 
+        # extra predict() parameters should be passed through correctly
+        with pytest.raises(AssertionError):
+            assert_eq(p1_raw, p1_first_iter)
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:712]:
```suggestion
            assert_eq(p1_raw, p1_first_iter_raw)
```

**Review by StrikerRUS — COMMENTED** [review]
Very nice catch! Just suggestion for more descriptive variable name below.
What about early stopping for prediction? Is it supported in Dask?
https://lightgbm.readthedocs.io/en/latest/Parameters.html#pred_early_stop

**Comment by jameslamb** [comment]
> What about early stopping for prediction? Is it supported in Dask?

I don't understand what "early stopping for prediction" actually means, can you explain it to me? The parameter descriptions at https://lightgbm.readthedocs.io/en/latest/Parameters.html#pred_early_stop are just short phrases using the same words as the parameter name (e.g. `pred_early_stop_margin = "the threshold of margin in early-stopping prediction"`), and I don't understand from the unit tests linked to in https://github.com/microsoft/LightGBM/pull/4399#pullrequestreview-690630824 what that functionality actually does.

I understand that early stopping for training means "stop the boosting process if performance on a validation set fails to improve", but I don't understand what early stopping means when you're generating predictions.

**Comment by StrikerRUS** [comment]
I guess it is something like "stop accumulating predictions of individual trees in final prediction if individual contributions are becoming insignificant".
Here is explanation from original author I just found: https://github.com/microsoft/LightGBM/pull/565#issuecomment-304803257.
Original PR: #550.

**Comment by jameslamb** [comment]
ahhh I see, interesting! Ok I can add calls with prediction early stopping to the tests in these PRs.

From #550, the tests you linked in https://github.com/microsoft/LightGBM/pull/4399#pullrequestreview-690630824 and my own investigation it seems that that parameter will only have an effect for classification objectives, so I'll only add it to the classifier tests.

**Comment by jameslamb** [comment]
Added a test with prediction early stopping in https://github.com/microsoft/LightGBM/pull/4399/commits/dccf44eb738aebef730d1bc46da0ea4f59b8cf9a

File: tests/python_package_test/test_dask.py [hunk 7]
```diff
@@ -272,6 +272,15 @@ def test_classifier(output, task, boosting_type, tree_learner, cluster):
...
         p1 = dask_classifier.predict(dX)
+        p1_raw = dask_classifier.predict(dX, raw_score=True).compute()
+        p1_first_iter_raw = dask_classifier.predict(dX, start_iteration=0, num_iteration=1, raw_score=True).compute()
+        p1_early_stop_raw = dask_classifier.predict(
+            dX,
+            pred_early_stop=True,
+            pred_early_stop_margin=1.0,
+            pred_early_stop_freq=2,
+            raw_score=True
+        )
```
Comment (StrikerRUS) [inline tests/python_package_test/test_dask.py:283]: Just curious: why does this particular line not ends with `.compute()`?

**Review by StrikerRUS — APPROVED** [review]
Thanks for the fix!

File: tests/python_package_test/test_dask.py (hunk 7 above)
Comment (jameslamb) [reply tests/python_package_test/test_dask.py:283]: just an oversight, there should be a `.compute()`. I've opened #4412 to add it.
